engine/kafka/src/main/scala/pl/touk/nussknacker/engine/kafka/validator/TopicsExistenceValidator.scala

```diff
@@ -0,0 +1,107 @@
+}
+
+case class CachedTopicsExistenceValidatorConfig(autoCreateFlagFetchCacheTtl: FiniteDuration, topicsFetchCacheTtl: FiniteDuration, adminClientTimeout: FiniteDuration)
+
+object CachedTopicsExistenceValidatorConfig {
+  val AutoCreateTopicPropertyName = "auto.create.topics.enable"
+  val TopicExistenceValidationEnabledPropertyName = "topic.existence.validation.enabled"
+  val ValidationEnabledDefault = false
+  val DefaultConfig = CachedTopicsExistenceValidatorConfig(
+    autoCreateFlagFetchCacheTtl = 5 minutes, topicsFetchCacheTtl = 30 seconds, adminClientTimeout = 500 millis
+  )
+}
+
+class CachedTopicsExistenceValidator(kafkaConfig: KafkaConfig, config: CachedTopicsExistenceValidatorConfig)
+  extends TopicsExistenceValidator with LazyLogging {
+
+  private lazy val autoCreateSettingCache = new SingleValueCache[Boolean](expireAfterAccess = None, expireAfterWrite = Some(config.autoCreateFlagFetchCacheTtl))
+  private lazy val topicListCache = new SingleValueCache[Set[String]](expireAfterAccess = None, expireAfterWrite = Some(config.topicsFetchCacheTtl))
+
+  private lazy val adminClientProperties = {
+    val properties = new Properties()
+    properties.setProperty("bootstrap.servers", kafkaConfig.kafkaAddress)
+    properties
+  }
+
+  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] = {
+    if (validationDisabled || isAutoCreateEnabled) {
+      Valid(topics)
+    } else {
+      val existingTopics = topicListCache.getOrCreate {
+        usingAdminClient {
+          _.listTopics(new ListTopicsOptions().timeoutMs(config.adminClientTimeout.toMillis.toInt))
+            .names().get().asScala.toSet
+        }
+      }
+      val notExistingTopics = topics.diff(existingTopics.toList)
+
+      if (notExistingTopics.isEmpty)
+        Valid(topics)
+      else
+        Invalid(new TopicExistenceValidationException(notExistingTopics))
+    }
+  }
+
+  def validationDisabled: Boolean = !kafkaConfig.kafkaEspProperties.flatMap(_.get(TopicExistenceValidationEnabledPropertyName))
+    .map(_.toBoolean).getOrElse(ValidationEnabledDefault)
+
+  private def isAutoCreateEnabled: Boolean = autoCreateSettingCache.getOrCreate {
+    val timeout = config.adminClientTimeout.toMillis.toInt
+    val id = usingAdminClient {
+      _.describeCluster(new DescribeClusterOptions().timeoutMs(timeout)).nodes().get().asScala.head.id().toString
+    }
+    usingAdminClient {
+      _
+        .describeConfigs(List(new ConfigResource(ConfigResource.Type.BROKER, id)).asJava,
+          new DescribeConfigsOptions().timeoutMs(config.adminClientTimeout.toMillis.toInt))
+        .values()
+        .values()
+        .asScala
+        .map(_.get())
+        .head.get(AutoCreateTopicPropertyName)
+        .value()
+        .toBoolean
+    }
+  }
+
+  private def usingAdminClient[T](adminClientOperation: Admin => T): T = {
```

Comment:
I wonder, maybe we can extract some of it to KafkaUtils?

Reply:
good point